### ISSUE TITLE
Blocking operations and aio MultiplexedConnection

### ISSUE THREAD

**Issue by mattklein123** [redis-rs/redis-rs#1236]
For fairly obvious reasons it isn't possible to share a MultiplexedConnection and use blocking operations (like XREAD with block), so you end up needing something like the existing deprecated basic aio Connection (or something specific like PubSub).

What is the recommended path forward here? I think having a basic connection is useful for reasons like this and in general custom connection pooling logic (deadpool redis uses MultiplexedConnection, but should probably just use the basic Connection). Is it reasonable to undeprecate Connection for this reason?

Separately (and I'm happy to open a different issue on this), I noticed that the lifetime of a MultiplexedConnection is not bound to the client AFAICT. Meaning, the connection driver is spawned onto the runtime. It would be useful to be able to close all existing connections or at least have some way of closing a multiplexed connection driver. Let me know if this is expected or not and whether I should open a different issue on this. 

Thank you for a great project!

**Comment by jaymell** [comment]
> For fairly obvious reasons it isn't possible to share a MultiplexedConnection and use blocking operations (like XREAD with block), so you end up needing something like the existing deprecated basic aio Connection (or something specific like PubSub).
> What is the recommended path forward here? I think having a basic connection is useful for reasons like this and in general custom connection pooling logic (deadpool redis uses MultiplexedConnection, but should probably just use the basic Connection). Is it reasonable to undeprecate Connection for this reason?

I have similar use cases in various projects I work with. My solution is to use connection pooling so that the connections aren't shared, [bb8](https://github.com/djc/bb8) in particular. We are definitely not going to undeprecate Connection, which is a security risk that should have been removed a long time ago. 

> Separately (and I'm happy to open a different issue on this), I noticed that the lifetime of a MultiplexedConnection is not bound to the client AFAICT. Meaning, the connection driver is spawned onto the runtime. It would be useful to be able to close all existing connections or at least have some way of closing a multiplexed connection driver. Let me know if this is expected or not and whether I should open a different issue on this.

I would probably recommend implementing your own wrapper type if you want to avoid cloning connections and/or want to track the state of all existing connections. I imagine that could be done without too much work, but I'm not sure it's something we want to support in the library. The ability to close all connections could be troublesome with cloneable connections.

**Comment by mattklein123** [comment]
> I have similar use cases in various projects I work with. My solution is to use connection pooling so that the connections aren't shared, [bb8](https://github.com/djc/bb8) in particular.

Yes we already use deadpool for this.

> We are definitely not going to undeprecate Connection, which is a security risk that should have been removed a long time ago.

Just for my understanding can you say more about this?

**Comment by mattklein123** [comment]
> I would probably recommend implementing your own wrapper type if you want to avoid cloning connections and/or want to track the state of all existing connections. I imagine that could be done without too much work, but I'm not sure it's something we want to support in the library. The ability to close all connections could be troublesome with cloneable connections.

The issue here is that the backing driver for MultiplexedConnection is spawned onto the runtime without any way to terminate it. Is there some way I am missing? I'm not sure how a wrapper could accomplish this.

**Comment by mattklein123** [comment]
> The connection won't persist forever if there's no more handles to it. Underlying MultiplexedConnection is a channel that will terminate if there are no more handles to the sending side of the channel.

I haven't debugged fully, but what I'm seeing is that if the connection is pending a blocking call, even if all handles are dropped, it sticks around until the call is completed. Optimally I would like a way to force close the connection in this case also.

**Comment by jaymell** [comment]
If you're using a connection pool, why are you concerned about blocking calls? In the crate I use, anyway, connections won't be shared when I check one out from the pool, so the blocking call is not a concern. This is the whole reason for using the connection pool.

**Comment by mattklein123** [comment]
> If you're using a connection pool, why are you concerned about blocking calls? In the crate I use, anyway, connections won't be shared when I check one out from the pool, so the blocking call is not a concern. This is the reason I use the connection pool.

Apologies I should have provided more context. This is mostly for certain testing scenarios around redis streams. The specific sequence that is complicated is something like this:
1) Start blocking xread
2) Drop single handle
3) Call remove consumer
4) Test goes on and does more stuff, xread gets unblocked returns, reruns xread, and then re-adds the consumer, before realizing the receiving handle is gone and it all drops.

And yes I understand that this is solved by XCLAIM, etc. but it would be nice to be able to force drop in certain cases. Optimally I would like some type of shutdown() which just force closes everything and fails any existing and further sends from handles.

**Comment by nihohit** [comment]
IMO this is a perfectly reasonable stance from a resource usage standpoint. If you have a socket, you want some deterministic way of terminating it. There's a way to do it today - instead of using `Client::**get**_multiplexed_...` use `Client::**create**_multiplexed_...`, which return both the connection and the function to spawn as a task. Then you can spawn the task and save the handle, and abort the task at your leisure. It's not an ergonomic solution, but it should work.

I'd like to have a more ergonomic solution, and I think that there are multiple ways of doing this - 

-  save a wrapper to a `Task` in an `Arc` and on the wrapper's `drop` abort the task, but I don't know how simple to implement this will be in a cross-runtime way.
-  add a signal or oneshot channel and in the spun-out task select on it and the current flow in order to terminate when the oneshot is called. not a big fan of adding more to the multiplexed connection - it's already much larger than the single sender it once was.
- in the task, selecting over the current loop and a periodic check that the receiver /stream is still open.

IMO this is worth pursuing (WDYT, @jaymell?), but none of these are simple changes, and I'm not sure that any of these will necessarilly work. If you'd like to explore this, I'd be happy to help, but I'm not sure I have the time to try and do this myself right now.

**Comment by jaymell** [comment]
> IMO this is worth pursuing (WDYT, @jaymell?), but none of these are simple changes, they'd need to come with significant testing in order to verify that we're not reintroducing the correctness issues of aio::Connection...

I'm not opposed to it but also don't feel like making this more ergonomic is particularly important. @mattklein123 Does the `create_` method meet your needs in your testing scenarios?

**Comment by mattklein123** [comment]
> I'm not opposed to it but also don't feel like making this more ergonomic is particularly important. @mattklein123 Does the create_ method meet your needs in your testing scenarios?

Yes this is a perfect workaround. Thank you!

**Comment by nihohit** [comment]
Leaving this out here for posterity:
I believe that this problem can be solved by the following struct. I'm not inclined to add this to the crate ATM, because I don't want to increase the maintenance load, but this could be considered for a future large-scale refactoring.
THIS WASN'T TESTED, USE WITH CAUTION!

```rs
struct Wrapper {
    conn: MultiplexedConnection,
    handle: tokio::task::JoinHandle<()>,
}

struct SharedConnection(Arc<Wrapper>);

impl Drop for Wrapper {
    fn drop(&mut self) {
        self.handle.abort();
    }
}

impl SharedConnection {
    fn new(conn: MultiplexedConnection, handle: tokio::task::JoinHandle<()>) -> Self {
        Self(Arc::new(Wrapper { conn, handle }))
    }

    // upside - this can be called with shared reference!
    async fn req_command<T: FromRedisValue>(&self, cmd: &Cmd) -> RedisResult<T> {
        cmd.query_async::<_, T>(&mut self.0.conn.clone()).await
    }
}

impl ConnectionLike for SharedConnection {
    fn req_packed_command<'a>(&'a mut self, cmd: &'a Cmd) -> RedisFuture<'a, Value> {
        let mut conn = self.0.conn.clone();
        (async move { conn.send_packed_command(cmd).await }).boxed()
    }

    fn req_packed_commands<'a>(
        &'a mut self,
        cmd: &'a crate::Pipeline,
        offset: usize,
        count: usize,
    ) -> RedisFuture<'a, Vec<Value>> {
        let mut conn = self.0.conn.clone();
        (async move { conn.send_packed_commands(cmd, offset, count).await }).boxed()
    }

    fn get_db(&self) -> i64 {
        self.0.conn.get_db()
    }
}

**Comment by mattklein123** [comment]
FWIW I wrote this for use with deadpool and it works fine: https://gist.github.com/mattklein123/3c5de708da24cc8d50919e774bb58b5b

**Comment by nihohit** [comment]
The fix was simpler than I expected: 
https://github.com/redis-rs/redis-rs/pull/1264

**Comment by louismlb** [comment]
> FWIW I wrote this for use with deadpool and it works fine: https://gist.github.com/mattklein123/3c5de708da24cc8d50919e774bb58b5b

Is your Drop implementation actually called by the deadpool code? If I'm correct deadpool puts back your WrapperConnection inside the pool when the Drop method on its `Object` is called. Testing your struct actually still gives me a timeout when trying to recycle a connection which was dropped while blocking.

**Comment by mattklein123** [comment]
> Is your Drop implementation actually called by the deadpool code @mattklein123 ?
If I'm correct deadpool puts back your WrappedConnection inside the pool when the Drop method on its managed::Object is called. Its own Drop method would thus not be called, and the driver never freed.
In fact testing your struct still seems to give me a timeout when trying to recycle a connection which was dropped while waiting on a blocking command.

Correct on its own it won't work, you have to also take the connection from the pool. I have external code which tracks drops for known blocking ops and does something like this: 
```
impl Drop for BlockingOpsImpl {
  fn drop(&mut self) {
    if !self.safe_to_recycle {
      log::debug!("dropped during blocking operation, not returning to pool");
      let _ = Object::take(self.connection.take().unwrap());
    }
  }
}
```

**Comment by louismlb** [comment]
Oh I guess that does it! Thank you so much!

**Comment by nihohit** [comment]
@louismlb @mattklein123 just released version 0.27.0 with a fix for this, so hopefully you can remove the extra code.
